calyx-backend/src/firrtl.rs

```diff
@@ -0,0 +1,181 @@
+//! FIRRTL backend for the Calyx compiler.
+//!
+//! Transforms an [`ir::Context`](crate::ir::Context) into a formatted string that represents a
+//! valid FIRRTL program.
+
+use crate::{traits::Backend, VerilogBackend};
+use calyx_ir::{self as ir};
+use calyx_utils::{CalyxResult, Error, OutputFile};
+use std::io;
+use std::time::Instant;
+
+pub(super) const SPACING: &str = "    ";
+
+/// Implements a simple FIRRTL backend. The backend only accepts Calyx programs with no control
+/// and no groups.
+#[derive(Default)]
+pub struct FirrtlBackend;
+
+impl Backend for FirrtlBackend {
+    fn name(&self) -> &'static str {
+        "firrtl"
+    }
+
+    fn link_externs(
+        _prog: &calyx_ir::Context,
+        _write: &mut calyx_utils::OutputFile,
+    ) -> calyx_utils::CalyxResult<()> {
+        Ok(()) // FIXME: Need to implement
+    }
+
+    fn validate(prog: &calyx_ir::Context) -> calyx_utils::CalyxResult<()> {
+        VerilogBackend::validate(prog) // FIXME: would this work if we wanted to check for the same things?
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let out = &mut file.get_write();
+        let comps = ctx.components.iter().try_for_each(|comp| {
+            // Time the generation of the component.
+            let time = Instant::now();
+            let out = emit_component(comp, out);
+            log::info!("Generated `{}` in {:?}", comp.name, time.elapsed());
+            out
+        });
+        comps.map_err(|err| {
+            let std::io::Error { .. } = err;
+            Error::write_error(format!(
+                "File not found: {}",
+                file.as_path_string()
+            ))
+        })
+    }
+}
+
+// TODO: Ask about the other backend configurations in verilog.rs and see if I need any of it
+fn emit_component<F: io::Write>(
+    comp: &ir::Component,
+    f: &mut F,
+) -> io::Result<()> {
+    writeln!(f, "circuit {}:", comp.name)?;
+    writeln!(f, "{}module {}:", SPACING, comp.name)?;
+
+    // Inputs and Outputs
+    let sig = comp.signature.borrow();
+    for (_idx, port_ref) in sig.ports.iter().enumerate() {
+        let port = port_ref.borrow();
+        let direction_string =
+        // NOTE: The signature port definitions are reversed inside the component.
+        match port.direction {
+            ir::Direction::Input => {"output"}
+            ir::Direction::Output => {"input"}
+            ir::Direction::Inout => {
+                panic!("Unexpected Inout port on Component: {}", port.name)
+            }
+        };
+        // FIXME: Hack to get clock declaration right. Should check for attribute name instead.
+        if port.name == "clk" {
```

Comment:
This should probably do what you're after
```suggestion
       port.has_attribute(ir::BoolAttr::Clk) {
```

Reply:
Thanks for helping me on my next TODO :smile: